torchao/experimental/quant_api.py

```diff
@@ -482,58 +487,139 @@ def quantize(self, model: nn.Module) -> nn.Module:
+        )
+
+    dtype_to_bit_width = {
+        torch.int1: 1,
+        torch.int2: 2,
+        torch.int3: 3,
+        torch.int4: 4,
+        torch.int5: 5,
+        torch.int6: 4,
+        torch.int7: 7,
+        torch.int8: 8,
+    }
+    if weight_dtype not in dtype_to_bit_width:
+        raise ValueError(
+            f"weight_dtype must be one of {list(dtype_to_bit_width.keys())}, got {weight_dtype}"
+        )
+    bit_width = dtype_to_bit_width[weight_dtype]
+    layout_arg = layout
 
     def apply(weight):
+        if isinstance(granularity, PerGroup):
+            group_size = granularity.group_size
+        elif isinstance(granularity, PerRow):
+            group_size = weight.shape[-1]
+        else:
+            raise ValueError(
+                f"granularity must be PerGroup or PerRow, got {granularity}"
+            )
+
         assert weight.shape[-1] % group_size == 0
-        assert weight.device == torch.device("cpu"), "Only CPU is supported"
-        use_hqq = False
-        layout = Linear8BitActXBitWeightLayout(
-            nbit=nbit, group_size=group_size, target=target
-        )
-        mapping_type = MappingType.ASYMMETRIC
-        eps = torch.finfo(torch.float32).eps
-        block_size = (1, group_size)
-        target_dtype = torch.int32
-        quant_min = -(1 << (nbit - 1))
-        quant_max = (1 << (nbit - 1)) - 1
-        zero_point_dtype = torch.int8
-        preserve_zero = has_weight_zeros
-        zero_point_domain = (
-            ZeroPointDomain.INT if has_weight_zeros else ZeroPointDomain.NONE
-        )
-        # Note: this works differently than other quantizers because the dynamic
-        # activation quantization is fused with the kernel/op (and static activation quantization
-        # is not supported).
-        return to_affine_quantized_intx(
+
+        layout = layout_arg
+        if isinstance(layout, PackedLinearInt8DynamicActivationIntxWeightLayout):
+            assert (
+                weight.device == torch.device("cpu")
+            ), "PackedLinearInt8DynamicActivationIntxWeightLayout requires weight.device=CPU"
+            assert (
+                weight.dtype == torch.float32
+            ), "PackedLinearInt8DynamicActivationIntxWeightLayout requires weight.dtype=float32"
+            assert (
+                act_mapping_type == MappingType.ASYMMETRIC
+            ), "PackedLinearInt8DynamicActivationIntxWeightLayout requires act_mapping_type=MappingType.ASYMMETRIC"
+            assert not layout.has_params_set(), "PackedLinearInt8DynamicActivationIntxWeightLayout params should not already be set"
+            layout = PackedLinearInt8DynamicActivationIntxWeightLayout(
+                bit_width=bit_width,
+                group_size=group_size,
+                has_weight_zeros=has_weight_zeros,
+            )
+
+        quant_min = -(1 << (bit_width - 1))
+        quant_max = (1 << (bit_width - 1)) - 1
```

Comment:
we also have utils here: https://github.com/pytorch/ao/blob/b3deb16e131fb9ba6039dbc7191bd3b4409ccc5f/torchao/quantization/quant_primitives.py#L168